### PR TITLE
Qmaps 1354 - show addresses in destination panel fields in case of latlon pois

### PR DESCRIPTION
**Pull request by xem**
## Description
Complete destination fields with latlon POI's addresses when it's possible
- when clicking on the map
- when restoring pois on page reload

## Why
It's more readable than GPS coordinates

## Screenshots
chelou2

### CODE REVIEW

File: src/panel/direction/DirectionForm.jsx [hunk 1]
```diff
@@ -11,15 +11,14 @@ export default class DirectionForm extends React.Component {
     destination: PropTypes.object,
     onChangeDirectionPoint: PropTypes.func.isRequired,
     onReversePoints: PropTypes.func.isRequired,
+    onEmptyOrigin: PropTypes.func.isRequired,
```
Comment (bbecquet) [inline src/panel/direction/DirectionForm.jsx]: These two functions are required… but unused. Looks like obsolete WIP code you forgot to remove. (And it reminds me we can add an ESLint rule to highlight that)

File: src/panel/direction/DirectionPanel.jsx [hunk 2]
```diff
@@ -206,13 +245,27 @@ export default class DirectionPanel extends React.Component {
...
       origin: previousState.destination,
       destination: previousState.origin,
+      originInputText: this.state.destinationInputText,
+      destinationInputText: this.state.originInputText,
       isDirty: true,
     }), this.update);
   }
 
-  changeDirectionPoint = (which, value) => {
+  emptyOrigin = () => {
```
Comment (bbecquet) [inline src/panel/direction/DirectionPanel.jsx]: According to the comment above, looks like you can delete these two functions.

File: src/panel/direction/DirectionPanel.jsx [hunk 3]
```diff
@@ -157,6 +177,25 @@ export default class DirectionPanel extends React.Component {
         fire('set_destination', destination);
       }
     }
+
+    let poiInfo;
```
Comment (bbecquet) [inline src/panel/direction/DirectionPanel.jsx]: By putting the reverse geocoding here, you postpone it to after the route is computed and displayed, whereas it could be done in parallel.
The two geocoding themselves could be done in parallel, but using two `await` make it in practice synchronized.
As a result, it's slower than needed.

This is also a bit strange to find these operations in a method called `computeRoutes` which explicitly tries to limit its scope to a single purpose. IMO it should be done in the `changeDirectionPoint`, so everything can be decoupled and can run more effectively.

File: src/panel/direction/DirectionPanel.jsx [hunk 4]
```diff
@@ -92,6 +95,16 @@ export default class DirectionPanel extends React.Component {
     Promise.all(poiRestorePromises).then(([ origin, destination ]) => {
       persistentPointState.origin = origin;
       persistentPointState.destination = destination;
+
+      // Show the raw POI name/latlon while waiting for computeRoutes to complete
+      if (origin) {
+        this.setState({ originInputText: origin.name });
```
Comment (bbecquet) [inline src/panel/direction/DirectionPanel.jsx]: Even if React can combine some operations behind the scene, it's better to regroup `setState` calls where it's possible, to limit useless re-renders.
Here, this can be grouped with the existing `setState` below.

**Comment by xem** [comment]
> 
> 
> Some comments, mostly about improving state management.
> 
> I also noticed a bug when testing:
> If I compute a route, then empty a field (with the X button), I can't click on the map to fill it again. Clicks are just ignored.

fixed :)

File: src/panel/direction/DirectionPanel.jsx (hunk 3 above)
Comment (xem) [reply src/panel/direction/DirectionPanel.jsx]: Hi,
I agree both requests must be made in parallel, I'm trying to fix it.
Regarding the place where the code should be, I initially chose computeRoutes because it seemed to be the only function called on field completion and on page reload, which is not the case for changeDirectionPoint (yet).
Maybe changeDirectionPoint only should compute addresses, and computeRoutes should call changeDirectionPoint, and if it does that, it could stop doing setStates directly... 🤔 
So I'm moving the code around hoping to find something more elegant, but I may need your help to confirm what is the nicest place/way to do it.

File: src/panel/direction/DirectionPanel.jsx [hunk 5]
```diff
@@ -92,6 +95,14 @@ export default class DirectionPanel extends React.Component {
     Promise.all(poiRestorePromises).then(([ origin, destination ]) => {
       persistentPointState.origin = origin;
       persistentPointState.destination = destination;
+
+      // Show the raw POI name/latlon while waiting for computeRoutes to complete
+      this.setState({
+        originInputText: origin ? origin.name : '',
+        destinationInputText: destination ? destination.name : '',
```
Comment (amatissart) [inline src/panel/direction/DirectionPanel.jsx]: To be consistent with the current implementation, addresses and streets should be rendered as labels (with city names)
```suggestion
        originInputText: origin ? origin.getInputValue() : '',
        destinationInputText: destination ? destination.getInputValue() : '',
```

File: src/panel/direction/DirectionPanel.jsx (hunk 3 above)
Comment (amatissart) [reply src/panel/direction/DirectionPanel.jsx]: Another side effect of calling Idunn here, is that new requests are executed on each compute (eg when another vehicle is selected). That should not be necessary.

File: src/panel/direction/DirectionPanel.jsx [hunk 6]
```diff
@@ -157,6 +175,25 @@ export default class DirectionPanel extends React.Component {
         fire('set_destination', destination);
       }
     }
+
+    let poiInfo;
+    if (origin) {
+      if (origin.type === 'latlon') {
+        poiInfo = await IdunnPoi.poiApiLoad(origin);
```
Comment (amatissart) [inline src/panel/direction/DirectionPanel.jsx]: Note that `IdunnPoi.poiApiLoad` returns an instance of `IdunnPoi` that could probably be used as `origin` or `destination` state directly.  
Then, the response would be persisted and would't require other requests on next computes.

File: src/panel/direction/DirectionPanel.jsx [hunk 7]
```diff
@@ -206,13 +259,26 @@ export default class DirectionPanel extends React.Component {
     this.setState(previousState => ({
       origin: previousState.destination,
       destination: previousState.origin,
+      originInputText: this.state.destinationInputText,
```
Comment (bbecquet) [inline src/panel/direction/DirectionPanel.jsx]: `setState` is a potentially asynchronous operation, that's why when you set values based on other state values, you use this function form which gives safe access to the current state through it's first argument. So it's safer to do `previousState.destinationInputText`, like it's done above.
Here is a link to the explanation in the React doc. https://fr.reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous

File: tests/integration/tests/direction.js [hunk 8]
```diff
@@ -91,7 +91,7 @@ test('destination', async () => {
   const directionStartInput = await page.evaluate(() =>
     document.getElementById('itinerary_input_origin').value
   );
-  expect(directionStartInput).toEqual('');
+  expect(directionStartInput).toEqual(' ');
```
Comment (bbecquet) [inline tests/integration/tests/direction.js]: Why a single space character?

File: src/panel/direction/DirectionPanel.jsx (hunk 7 above)
Comment (xem) [reply src/panel/direction/DirectionPanel.jsx]: k, can tou confirm me that I should also use previousState on lines 105 & 110 ?

File: src/panel/direction/DirectionPanel.jsx [hunk 9]
```diff
@@ -206,13 +259,26 @@ export default class DirectionPanel extends React.Component {
...
+    persistentPointState[which] = point;
+    this.setState({
+      [which]: point,
+      isDirty: true,
+      [which + 'InputText']: value || '',
+    }, this.update);
+
+    // Retrieve addresses
+    if (which === 'origin') {
+      this.getAddresses(persistentPointState.origin, null);
```
Comment (bbecquet) [inline src/panel/direction/DirectionPanel.jsx]: I still don't get the point of the `getAddresses` function taking the two points. Among other things, it forces you to make these calls with `null` which look unnatural. As the action on both points is exactly the same, why not having a simpler function?

File: src/panel/direction/DirectionPanel.jsx [hunk 10]
```diff
@@ -80,6 +83,35 @@ export default class DirectionPanel extends React.Component {
...
+          ? IdunnPoi.poiApiLoad(origin)
+          : Promise.resolve(origin)
+        : null,
+      destination
+        ? destination.type === 'latlon'
+          ? IdunnPoi.poiApiLoad(destination)
+          : Promise.resolve(destination)
+        : null,
+    ];
+    Promise.all(inputTextPromises).then(([ origin, destination ]) => {
```
Comment (bbecquet) [inline src/panel/direction/DirectionPanel.jsx]: If I'm not wrong, there is no need for this forced synchronization. Any of two points text can be updated independantly from the other, as soon as we get the response.

File: src/panel/direction/DirectionPanel.jsx [hunk 11]
```diff
@@ -80,6 +83,35 @@ export default class DirectionPanel extends React.Component {
     document.body.classList.remove('directions-open');
   }
 
+  getAddresses = (origin, destination) => {
+    const inputTextPromises = [
+      origin
```
Comment (bbecquet) [inline src/panel/direction/DirectionPanel.jsx]: We should find a way to avoid these unclear nested ternaries.
For example:
```js
origin && origin.type === 'latlon' ? IdunnPoi.poiApiLoad(origin) : origin
```
And sorry about the `Promise.resolve` that I used elsewhere, you can remove it if you want.
I thought it was needed but it can be omitted with `Promise.all`, any non-Promise value is considered as a Promise resolving with this value.

File: src/panel/direction/DirectionPanel.jsx [hunk 12]
```diff
@@ -92,6 +124,16 @@ export default class DirectionPanel extends React.Component {
     Promise.all(poiRestorePromises).then(([ origin, destination ]) => {
       persistentPointState.origin = origin;
       persistentPointState.destination = destination;
+
+      // Show the raw POI name/latlon while waiting for computeRoutes to complete
+      if (origin || destination) {
+        this.setState({
+          originInputText: ' ',
```
Comment (bbecquet) [inline src/panel/direction/DirectionPanel.jsx]: Why a single space character instead of an empty string like it's already done in the initial state?

File: src/panel/direction/DirectionPanel.jsx (hunk 9 above)
Comment (xem) [reply src/panel/direction/DirectionPanel.jsx]: I agreen but if I make this function perform only one request, I don't know how to call it twice without making the two function calls (and the two API calls) synchronous...

File: src/panel/direction/DirectionPanel.jsx (hunk 12 above)
Comment (xem) [reply src/panel/direction/DirectionPanel.jsx]: to avoid opening the suggest while the address is loading (didn't find how to disable it yet)

File: src/panel/direction/DirectionPanel.jsx (hunk 9 above)
Comment (bbecquet) [reply src/panel/direction/DirectionPanel.jsx]: That's my point, I don't see why you need to synchronize the calls. Is there a UX reason?
Because otherwise, it's like remote work: the less synchronous stuff the better :smile:

File: src/panel/direction/DirectionPanel.jsx (hunk 12 above)
Comment (bbecquet) [reply src/panel/direction/DirectionPanel.jsx]: Maybe a less hackish solution would be a dedicated state value, indicating that the form is still initializing and preventing the suggest to appear?

File: src/panel/direction/DirectionPanel.jsx [hunk 13]
```diff
@@ -120,8 +145,15 @@ export default class DirectionPanel extends React.Component {
...
-      this.setState({ isDirty: false, isLoading: true, error: 0, routes: [] });
+      this.setState({
+        isDirty: false,
+        isLoading: true,
+        error: 0,
+        routes: [],
+      });
       const currentQueryId = ++this.lastQueryId;
+      fire('set_origin', origin);
+      fire('set_destination', destination);
```
Comment (amatissart) [inline src/panel/direction/DirectionPanel.jsx]: Why is this change required ?

Comment (xem) [reply src/panel/direction/DirectionPanel.jsx]: the idea was to show the start/end markers without waiting for the route to be computed (which can take a few seconds). If I remove these lines, and compute a route between two points clicked on the map, the last marker I clicked appears at the same time as the route, which is too late imo...

Comment (xem) [reply src/panel/direction/DirectionPanel.jsx]: ...Do you want me to make sure any preexisting marker is cleared before that? The way I did it, one of the markers may get doubled (harmlessly)

Comment (amatissart) [reply src/panel/direction/DirectionPanel.jsx]: I am not really worried about the markers, and I agree that the markers should be visible as soon as possible.  

But a side-effect of this change is that `set_origin` and `set_destination` will call `fitMap` before the route is computed.
So selecting the origin or destination that will trigger the compute, will actually cause the map position to jump twice, before the route is visible.

Comment (xem) [reply src/panel/direction/DirectionPanel.jsx]: OH, indeed. That's wrong. I guess I could add a fitMap = false parameter to setOrigin/Destination... is it ok for you?

Comment (amatissart) [reply src/panel/direction/DirectionPanel.jsx]: Yes, it should be ok :+1:

Comment (xem) [reply src/panel/direction/DirectionPanel.jsx]: ... but wait, I don't see anything that resembles fitMap() in these functions or in the functions called by them! 🤔 all they do is create a marker and add it to the map and in routeMarkers...

Comment (xem) [reply src/panel/direction/DirectionPanel.jsx]: actually I removed the fitMap entirely from set_origin and set_destination as there were zero cases where the fitMap was needed when calling them.
FitMap is only necessary in one case: when only one point/field is restored from the url, so I added the fitMap in this special case.
